File: java/yass-base/main/ch/softappeal/yass/serialize/fast/BaseTypeHandlers.java
package ch.softappeal.yass.serialize.fast;

import ch.softappeal.yass.serialize.Reader;
import ch.softappeal.yass.serialize.Utf8;
import ch.softappeal.yass.serialize.Writer;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.Instant;
import java.util.Arrays;
import java.util.Date;

public final class BaseTypeHandlers {

    private BaseTypeHandlers() {
        // disable
    }

    public static final BaseTypeHandler<Boolean> BOOLEAN = new BaseTypeHandler<>(Boolean.class) {
        @Override public Boolean read(final Reader reader) throws Exception {
            return reader.readByte() != 0;
        }
        @Override public void write(final Boolean value, final Writer writer) throws Exception {
            writer.writeByte(value ? (byte)1 : (byte)0);
        }
    };

    public static final BaseTypeHandler<Byte> BYTE = new BaseTypeHandler<>(Byte.class) {
        @Override public Byte read(final Reader reader) throws Exception {
            return reader.readByte();
        }
        @Override public void write(final Byte value, final Writer writer) throws Exception {
            writer.writeByte(value);
        }
    };

    public static final BaseTypeHandler<Short> SHORT = new BaseTypeHandler<>(Short.class) {
        @Override public Short read(final Reader reader) throws Exception {
            return (short)reader.readZigZagInt();
        }
        @Override public void write(final Short value, final Writer writer) throws Exception {
            writer.writeZigZagInt(value);
        }
    };

    public static final BaseTypeHandler<Integer> INTEGER = new BaseTypeHandler<>(Integer.class) {
        @Override public Integer read(final Reader reader) throws Exception {
            return reader.readZigZagInt();
        }
        @Override public void write(final Integer value, final Writer writer) throws Exception {
            writer.writeZigZagInt(value);
        }
    };

    public static final BaseTypeHandler<Long> LONG = new BaseTypeHandler<>(Long.class) {
        @Override public Long read(final Reader reader) throws Exception {
            return reader.readZigZagLong();
        }
        @Override public void write(final Long value, final Writer writer) throws Exception {
            writer.writeZigZagLong(value);
        }
    };

    public static final BaseTypeHandler<Character> CHARACTER = new BaseTypeHandler<>(Character.class) {
        @Override public Character read(final Reader reader) throws Exception {
            return reader.readChar();
        }
        @Override public void write(final Character value, final Writer writer) throws Exception {
            writer.writeChar(value);
        }
    };

    public static final BaseTypeHandler<Float> FLOAT = new BaseTypeHandler<>(Float.class) {
        @Override public Float read(final Reader reader) throws Exception {
            return reader.readFloat();
        }
        @Override public void write(final Float value, final Writer writer) throws Exception {
            writer.writeFloat(value);
        }
    };

    public static final BaseTypeHandler<Double> DOUBLE = new BaseTypeHandler<>(Double.class) {
        @Override public Double read(final Reader reader) throws Exception {
            return reader.readDouble();
        }
        @Override public void write(final Double value, final Writer writer) throws Exception {
            writer.writeDouble(value);
        }
    };

    public static final BaseTypeHandler<boolean[]> BOOLEAN_ARRAY = new BaseTypeHandler<>(boolean[].class) {
        @Override public boolean[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            boolean[] value = new boolean[Math.min(length, 128)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = reader.readByte() != 0;
            }
            return value;
        }
        @Override public void write(final boolean[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final boolean v : value) {
                writer.writeByte(v ? (byte)1 : (byte)0);
            }
        }
    };

    public static final BaseTypeHandler<byte[]> BYTE_ARRAY = new BaseTypeHandler<>(byte[].class) {
        @Override public byte[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            byte[] value = new byte[Math.min(length, 128)];
            int i = 0;
            while (i < length) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                final int l = value.length - i;
                reader.readBytes(value, i, l);
                i += l;
            }
            return value;
        }
        @Override public void write(final byte[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            writer.writeBytes(value);
        }
    };

    public static final BaseTypeHandler<short[]> SHORT_ARRAY = new BaseTypeHandler<>(short[].class) {
        @Override public short[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            short[] value = new short[Math.min(length, 64)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = (short)reader.readZigZagInt();
            }
            return value;
        }
        @Override public void write(final short[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final short v : value) {
                writer.writeZigZagInt(v);
            }
        }
    };

    public static final BaseTypeHandler<int[]> INTEGER_ARRAY = new BaseTypeHandler<>(int[].class) {
        @Override public int[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            int[] value = new int[Math.min(length, 32)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = reader.readZigZagInt();
            }
            return value;
        }
        @Override public void write(final int[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final int v : value) {
                writer.writeZigZagInt(v);
            }
        }
    };

    public static final BaseTypeHandler<long[]> LONG_ARRAY = new BaseTypeHandler<>(long[].class) {
        @Override public long[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            long[] value = new long[Math.min(length, 16)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = reader.readZigZagLong();
            }
            return value;
        }
        @Override public void write(final long[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final long v : value) {
                writer.writeZigZagLong(v);
            }
        }
    };

    public static final BaseTypeHandler<char[]> CHARACTER_ARRAY = new BaseTypeHandler<>(char[].class) {
        @Override public char[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            char[] value = new char[Math.min(length, 64)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = reader.readChar();
            }
            return value;
        }
        @Override public void write(final char[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final char v : value) {
                writer.writeChar(v);
            }
        }
    };

    public static final BaseTypeHandler<float[]> FLOAT_ARRAY = new BaseTypeHandler<>(float[].class) {
        @Override public float[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            float[] value = new float[Math.min(length, 32)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = reader.readFloat();
            }
            return value;
        }
        @Override public void write(final float[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final float v : value) {
                writer.writeFloat(v);
            }
        }
    };

    public static final BaseTypeHandler<double[]> DOUBLE_ARRAY = new BaseTypeHandler<>(double[].class) {
        @Override public double[] read(final Reader reader) throws Exception {
            final int length = reader.readVarInt();
            double[] value = new double[Math.min(length, 16)];
            for (int i = 0; i < length; i++) {
                if (i >= value.length) {
                    value = Arrays.copyOf(value, Math.min(length, 2 * value.length)); // note: prevents out-of-memory attack
                }
                value[i] = reader.readDouble();
            }
            return value;
        }
        @Override public void write(final double[] value, final Writer writer) throws Exception {
            writer.writeVarInt(value.length);
            for (final double v : value) {
                writer.writeDouble(v);
            }
        }
    };

    public static final BaseTypeHandler<String> STRING = new BaseTypeHandler<>(String.class) {
        @Override public String read(final Reader reader) throws Exception {
            return Utf8.string(BYTE_ARRAY.read(reader));
        }
        @Override public void write(final String value, final Writer writer) throws Exception {
            BYTE_ARRAY.write(Utf8.bytes(value), writer);
        }
    };

    public static final BaseTypeHandler<Date> DATE = new BaseTypeHandler<>(Date.class) {
        @Override public Date read(final Reader reader) throws Exception {
            return new Date(LONG.read(reader));
        }
        @Override public void write(final Date value, final Writer writer) throws Exception {
            LONG.write(value.getTime(), writer);
        }
    };

    public static final BaseTypeHandler<Instant> INSTANT = new BaseTypeHandler<>(Instant.class) {
        @Override public Instant read(final Reader reader) throws Exception {
            return Instant.ofEpochSecond(LONG.read(reader), reader.readVarInt());
        }
        @Override public void write(final Instant value, final Writer writer) throws Exception {
            LONG.write(value.getEpochSecond(), writer);
            writer.writeVarInt(value.getNano());
        }
    };

    public static final BaseTypeHandler<BigInteger> BIGINTEGER = new BaseTypeHandler<>(BigInteger.class) {
        @Override public BigInteger read(final Reader reader) throws Exception {
            return new BigInteger(BYTE_ARRAY.read(reader));
        }
        @Override public void write(final BigInteger value, final Writer writer) throws Exception {
            BYTE_ARRAY.write(value.toByteArray(), writer);
        }
    };

    public static final BaseTypeHandler<BigDecimal> BIGDECIMAL = new BaseTypeHandler<>(BigDecimal.class) {
        @Override public BigDecimal read(final Reader reader) throws Exception {
            return new BigDecimal(BIGINTEGER.read(reader), INTEGER.read(reader));
        }
        @Override public void write(final BigDecimal value, final Writer writer) throws Exception {
            BIGINTEGER.write(value.unscaledValue(), writer);
            INTEGER.write(value.scale(), writer);
        }
    };

}

File: kotlin/yass/main/ch/softappeal/yass/serialize/fast/BaseTypeHandlers.kt
@file:JvmName("Kt")
@file:JvmMultifileClass

package ch.softappeal.yass.serialize.fast

import ch.softappeal.yass.serialize.Reader
import ch.softappeal.yass.serialize.Writer
import ch.softappeal.yass.serialize.utf8toBytes
import ch.softappeal.yass.serialize.utf8toString
import java.math.BigDecimal
import java.math.BigInteger
import java.time.Instant
import java.util.Arrays
import java.util.Date

val BTH_BOOLEAN = object : BaseTypeHandler<Boolean>(Boolean::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readByte().toInt() != 0
    override fun write(writer: Writer, value: Boolean) = writer.writeByte(if (value) 1.toByte() else 0.toByte())
}

val BTH_BYTE = object : BaseTypeHandler<Byte>(Byte::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readByte()
    override fun write(writer: Writer, value: Byte) = writer.writeByte(value)
}

val BTH_SHORT = object : BaseTypeHandler<Short>(Short::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readZigZagInt().toShort()
    override fun write(writer: Writer, value: Short) = writer.writeZigZagInt(value.toInt())
}

val BTH_INTEGER = object : BaseTypeHandler<Int>(Int::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readZigZagInt()
    override fun write(writer: Writer, value: Int) = writer.writeZigZagInt(value)
}

val BTH_LONG = object : BaseTypeHandler<Long>(Long::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readZigZagLong()
    override fun write(writer: Writer, value: Long) = writer.writeZigZagLong(value)
}

val BTH_CHARACTER = object : BaseTypeHandler<Char>(Char::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readChar()
    override fun write(writer: Writer, value: Char) = writer.writeChar(value)
}

val BTH_FLOAT = object : BaseTypeHandler<Float>(Float::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readFloat()
    override fun write(writer: Writer, value: Float) = writer.writeFloat(value)
}

val BTH_DOUBLE = object : BaseTypeHandler<Double>(Double::class.javaObjectType) {
    override fun read(reader: Reader) = reader.readDouble()
    override fun write(writer: Writer, value: Double) = writer.writeDouble(value)
}

val BTH_BOOLEAN_ARRAY = object : BaseTypeHandler<BooleanArray>(BooleanArray::class.java) {
    override fun read(reader: Reader): BooleanArray {
        val length = reader.readVarInt()
        var value = BooleanArray(Math.min(length, 128))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readByte().toInt() != 0
        }
        return value
    }

    override fun write(writer: Writer, value: BooleanArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeByte(if (v) 1.toByte() else 0.toByte())
    }
}

val BTH_BYTE_ARRAY = object : BaseTypeHandler<ByteArray>(ByteArray::class.java) {
    override fun read(reader: Reader): ByteArray {
        val length = reader.readVarInt()
        var value = ByteArray(Math.min(length, 128))
        var i = 0
        while (i < length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            val l = value.size - i
            reader.readBytes(value, i, l)
            i += l
        }
        return value
    }

    override fun write(writer: Writer, value: ByteArray) {
        writer.writeVarInt(value.size)
        writer.writeBytes(value)
    }
}

val BTH_SHORT_ARRAY = object : BaseTypeHandler<ShortArray>(ShortArray::class.java) {
    override fun read(reader: Reader): ShortArray {
        val length = reader.readVarInt()
        var value = ShortArray(Math.min(length, 64))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readZigZagInt().toShort()
        }
        return value
    }

    override fun write(writer: Writer, value: ShortArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeZigZagInt(v.toInt())
    }
}

val BTH_INTEGER_ARRAY = object : BaseTypeHandler<IntArray>(IntArray::class.java) {
    override fun read(reader: Reader): IntArray {
        val length = reader.readVarInt()
        var value = IntArray(Math.min(length, 32))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readZigZagInt()
        }
        return value
    }

    override fun write(writer: Writer, value: IntArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeZigZagInt(v)
    }
}

val BTH_LONG_ARRAY = object : BaseTypeHandler<LongArray>(LongArray::class.java) {
    override fun read(reader: Reader): LongArray {
        val length = reader.readVarInt()
        var value = LongArray(Math.min(length, 16))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readZigZagLong()
        }
        return value
    }

    override fun write(writer: Writer, value: LongArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeZigZagLong(v)
    }
}

val BTH_CHARACTER_ARRAY = object : BaseTypeHandler<CharArray>(CharArray::class.java) {
    override fun read(reader: Reader): CharArray {
        val length = reader.readVarInt()
        var value = CharArray(Math.min(length, 64))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readChar()
        }
        return value
    }

    override fun write(writer: Writer, value: CharArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeChar(v)
    }
}

val BTH_FLOAT_ARRAY = object : BaseTypeHandler<FloatArray>(FloatArray::class.java) {
    override fun read(reader: Reader): FloatArray {
        val length = reader.readVarInt()
        var value = FloatArray(Math.min(length, 32))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readFloat()
        }
        return value
    }

    override fun write(writer: Writer, value: FloatArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeFloat(v)
    }
}

val BTH_DOUBLE_ARRAY = object : BaseTypeHandler<DoubleArray>(DoubleArray::class.java) {
    override fun read(reader: Reader): DoubleArray {
        val length = reader.readVarInt()
        var value = DoubleArray(Math.min(length, 16))
        for (i in 0 until length) {
            if (i >= value.size) value = Arrays.copyOf(value, Math.min(length, 2 * value.size))
            value[i] = reader.readDouble()
        }
        return value
    }

    override fun write(writer: Writer, value: DoubleArray) {
        writer.writeVarInt(value.size)
        for (v in value) writer.writeDouble(v)
    }
}

val BTH_STRING = object : BaseTypeHandler<String>(String::class.java) {
    override fun read(reader: Reader) = utf8toString(BTH_BYTE_ARRAY.read(reader))
    override fun write(writer: Writer, value: String) = BTH_BYTE_ARRAY.write(writer, utf8toBytes(value))
}

val BTH_DATE = object : BaseTypeHandler<Date>(Date::class.java) {
    override fun read(reader: Reader) = Date(BTH_LONG.read(reader))
    override fun write(writer: Writer, value: Date) = BTH_LONG.write(writer, value.time)
}

val BTH_INSTANT = object : BaseTypeHandler<Instant>(Instant::class.java) {
    override fun read(reader: Reader) = Instant.ofEpochSecond(BTH_LONG.read(reader), reader.readVarInt().toLong())
    override fun write(writer: Writer, value: Instant) {
        BTH_LONG.write(writer, value.epochSecond)
        writer.writeVarInt(value.nano)
    }
}

val BTH_BIGINTEGER = object : BaseTypeHandler<BigInteger>(BigInteger::class.java) {
    override fun read(reader: Reader) = BigInteger(BTH_BYTE_ARRAY.read(reader))
    override fun write(writer: Writer, value: BigInteger) = BTH_BYTE_ARRAY.write(writer, value.toByteArray())
}

val BTH_BIGDECIMAL = object : BaseTypeHandler<BigDecimal>(BigDecimal::class.java) {
    override fun read(reader: Reader) = BigDecimal(BTH_BIGINTEGER.read(reader), BTH_INTEGER.read(reader))
    override fun write(writer: Writer, value: BigDecimal) {
        BTH_BIGINTEGER.write(writer, value.unscaledValue())
        BTH_INTEGER.write(writer, value.scale())
    }
}
